paimon-core/src/main/java/org/apache/paimon/operation/FileDeletionBase.java

```diff
@@ -227,32 +222,17 @@ protected Iterable<ManifestEntry> tryReadDataManifestEntries(Snapshot snapshot)
         return readManifestEntries(manifestFileMetas);
     }
 
-    protected Iterable<ManifestEntry> readManifestEntries(
-            List<ManifestFileMeta> manifestFileMetas) {
-        Queue<String> files =
-                manifestFileMetas.stream()
-                        .map(ManifestFileMeta::fileName)
-                        .collect(Collectors.toCollection(LinkedList::new));
-        return Iterables.concat(
-                (Iterable<Iterable<ManifestEntry>>)
-                        () ->
-                                new Iterator<Iterable<ManifestEntry>>() {
-                                    @Override
-                                    public boolean hasNext() {
-                                        return files.size() > 0;
-                                    }
-
-                                    @Override
-                                    public Iterable<ManifestEntry> next() {
-                                        String file = files.poll();
-                                        try {
-                                            return manifestFile.read(file);
-                                        } catch (Exception e) {
-                                            LOG.warn("Failed to read manifest file " + file, e);
-                                            return Collections.emptyList();
-                                        }
-                                    }
-                                });
+    /** Cancel reading when any file is failed to read. */
+    private List<ManifestEntry> readManifestEntries(List<ManifestFileMeta> manifestFileMetas) {
+        try {
+            return manifestFileMetas.stream()
```

Comment:
Here you break the lazy load for entries.

You can just cancel in the outer.

Reply:
Refactored. 
1. for data skipping, if the read failed, we should skip the cleaning because skipping set is not complete, we may delete used data file;
2. if we should delete the data file in any case (like `deleteAddedDataFiles`),  we can just ignore unavailable data file.